### PR TITLE
Streams finish properly after they have ended.

### LINKED ISSUE #66: Handle .intoPromise() and .intoCallback() after stream has ended
Currently, if a stream has finished or errored already, and .intoPromise() or .intoCallback() are called, the promise never resolves/callback is never called.  This should do the expected of returning a resolved/rejected promise or calling the callback immediately.

### PR DESCRIPTION
**Pull request by crowelch**
fixes #66

### CODE REVIEW

File: lib/mixins/_writable.js [hunk 1]
```diff
@@ -49,27 +59,40 @@ _Writable.prototype.firstFinish = function(handler, ignoreCloseAsFinish) {
 		// (not in addition to) 'finish'
 		this.once('close', handleEvent);
 	}
+
+	if(!!this._writableState.zEnded) {
```
Comment (crispy1989) [inline lib/mixins/_writable.js]: No need for the !!

File: lib/mixins/_writable.js [hunk 2]
```diff
@@ -49,27 +59,40 @@ _Writable.prototype.firstFinish = function(handler, ignoreCloseAsFinish) {
...
-		calledCb = true;
-		cb.call(this);
-	});
+_Writable.prototype._intoCallback = function(cb, error) {
+	cb.call(this, error);
 	return this;
 };
 
+_Writable.prototype.intoCallback = function(cb) {
+	if(!!this._writableState.zError) {
```
Comment (crispy1989) [inline lib/mixins/_writable.js]: No need for !!

File: lib/mixins/_writable.js [hunk 3]
```diff
@@ -49,27 +59,40 @@ _Writable.prototype.firstFinish = function(handler, ignoreCloseAsFinish) {
...
-		if(calledCb) return;
-		calledCb = true;
-		cb.call(this, error);
-	});
-	this.firstFinish(function() {
-		if(calledCb) return;
-		calledCb = true;
-		cb.call(this);
-	});
+_Writable.prototype._intoCallback = function(cb, error) {
```
Comment (crispy1989) [inline lib/mixins/_writable.js]: No need for this.

File: lib/mixins/_writable.js [hunk 4]
```diff
@@ -49,27 +59,40 @@ _Writable.prototype.firstFinish = function(handler, ignoreCloseAsFinish) {
...
+	cb.call(this, error);
 	return this;
 };
 
+_Writable.prototype.intoCallback = function(cb) {
+	if(!!this._writableState.zError) {
+		this._intoCallback(cb, this._writableState.zError);
```
Comment (crispy1989) [inline lib/mixins/_writable.js]: Get rid of helper.  Also be sure to return this .

File: lib/mixins/_writable.js [hunk 5]
```diff
@@ -25,6 +27,14 @@ function _Writable() {
 			return src !== source;
 		});
 	});
+
+	this.once('finish', function() {
```
Comment (crispy1989) [inline lib/mixins/_writable.js]: Use the builtin helpers for this (.firstFinish, .firstError)

File: lib/mixins/_writable.js (hunk 1 above)
Comment (crispy1989) [inline lib/mixins/_writable.js]: You're handling this here in firstFinish() but not in firstUnignoredError() .  Be consistent.

**Comment by crowelch** [comment]
boop

File: lib/mixins/_writable.js [hunk 6]
```diff
@@ -49,24 +59,35 @@ _Writable.prototype.firstFinish = function(handler, ignoreCloseAsFinish) {
...
 	}
+
+	if(this._writableState.zEnded) {
+		handleEvent();
+	}
 };
 
 _Writable.prototype.firstUnignoredError = function(handler) {
+	if(this._writableState.zEnded) {
+		handler();
```
Comment (crispy1989) [inline lib/mixins/_writable.js]: Can be called twice?

**Pull request merged**